src/web/middleware/rejectua.rs

```diff
@@ -0,0 +1,118 @@
+#![allow(clippy::type_complexity)]
+use std::task::{Context, Poll};
+
+use actix_web::{
+    dev::{Service, ServiceRequest, ServiceResponse, Transform},
+    http::header::USER_AGENT,
+    Error, HttpResponse,
+};
+use futures::future::{self, Either, Ready};
+use lazy_static::lazy_static;
+use regex::Regex;
+
+use crate::server::{metrics::Metrics, ServerState};
+
+lazy_static! {
+    // e.g. "Firefox-iOS-Sync/18.0b1 (iPhone; iPhone OS 13.2.2) (Fennec (synctesting))"
+    // https://github.com/mozilla-mobile/firefox-ios/blob/v19.x/Shared/UserAgent.swift#L12
+    static ref IOS_UA_REGEX: Regex = Regex::new(
+        r"(?x)
+^
+Firefox-iOS-Sync/
+(?P<major>[0-9]+)\.[.0-9]+    # <appVersion-major>.<appVersion-minor-etc>
+b.*                           # b<buildNumber>
+\s\(.+                        #  (<deviceModel>
+;\siPhone\sOS                 # ; iPhone OS
+\s.+\)                        #  <systemVersion>)
+\s\(.*\)                      #  (<displayName>)
+$
+"
+    )
+    .unwrap();
+}
+
+#[derive(Debug, Default)]
+pub struct RejectUA;
+
+impl<S, B> Transform<S> for RejectUA
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type InitError = ();
+    type Transform = RejectUAMiddleware<S>;
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        future::ok(RejectUAMiddleware { service })
+    }
+}
+
+pub struct RejectUAMiddleware<S> {
+    service: S,
+}
+
+impl<S, B> Service for RejectUAMiddleware<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = Either<Ready<Result<Self::Response, Self::Error>>, S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, sreq: ServiceRequest) -> Self::Future {
+        match sreq.headers().get(USER_AGENT) {
+            Some(header) if header.to_str().map_or(false, should_reject) => {
+                let state = match &sreq.app_data::<ServerState>() {
+                    Some(v) => v.clone(),
+                    None => {
+                        return Either::Left(future::ok(
+                            sreq.into_response(
+                                HttpResponse::InternalServerError()
+                                    .body("Err: No State".to_owned())
+                                    .into_body(),
+                            ),
+                        ))
+                    }
+                };
+                debug!("Rejecting User-Agent: {:?}", header);
```

Comment:
Tempted to ask if we could metric this. It'd be nice to see if the number drops far enough we can ignore it.

Reply:
It is metric'd! Spoiler alert it's on the next line

Reply:
how the hell did I miss that? sigh. 🤦‍♂